File: algorithms/sort/merge/python/merge.py
"""
>>> a = [1, 2, 3, 4, 5]
>>> merge_sort(a)
[1, 2, 3, 4, 5]
>>> a = [4, 2, 5, 1, 3]
>>> merge_sort(a)
[1, 2, 3, 4, 5]
"""


def merge_sort(a):
    if len(a) == 1:
        return a
    middle = int(len(a) / 2)
    sorted_1 = merge_sort(a[:middle])
    sorted_2 = merge_sort(a[middle:])
    return merge(sorted_1, sorted_2)


def merge(a1, a2):
   i, j = 0, 0
   result = []
   while i < len(a1) and j < len(a2):
       if a1[i] < a2[j]:
           result.append(a1[i])
           i += 1
       else:
           result.append(a2[j])
           j += 1
   while i < len(a1):
       result.append(a1[i])
       i += 1
   while j < len(a2):
       result.append(a2[j])
       j += 1
   return result


if __name__ == "__main__":
    import doctest
    doctest.testmod()

File: algorithms/sort/merge/C/merge.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void merge(int *arr, int begin, int middle, int end) {
    int i=begin; int j=middle+1; int k=0;
    int length = end-begin+1;
    int *temp;

    temp = (int*) malloc(length * sizeof(int));

    if (temp != NULL) {
        while (i <= middle && j <= end) {
            if (arr[i] < arr[j])
                temp[k++] = arr[i++];
            else
                temp[k++] = arr[j++];
        }

        while (i <= middle)
            temp[k++] = arr[i++];
        while (j <= end)
            temp[k++] = arr[j++];

        for (i=0, j=begin; i < length; i++, j++)
            arr[j] = temp[i];

        free(temp);
    }
}

void mergeSort(int *arr, int begin, int end) {
    if (begin < end) {
        int middle = floor((begin + end) / 2);
        mergeSort(arr, begin, middle);
        mergeSort(arr, middle+1, end);
        merge(arr, begin, middle, end);
    }
}

int main() {
    int n; int *arr;
    scanf("%d", &n);

    arr = (int*) malloc((n + 1) * sizeof(int));
    for (int i = 0; i < n; i++) {
        int temp = 0;
        scanf("%d", &temp);
        arr[i] = temp;
    }

    mergeSort(arr, 0, n-1);

    for (int i=0; i<n; i++)
        printf("%d ", arr[i]);

    return 0;
}
